### PR TITLE
CRM-20313,CRM-19357: Add index to civicrm_activity.status_id and unique index to civicrm_entity_financial_account table

### PR DESCRIPTION
**Pull request by Edzelopez**
* [CRM-20313: Add index to civicrm_activity.status_id](https://issues.civicrm.org/jira/browse/CRM-20313)

---

 * [CRM-19357: Add unique constraint on civicrm_entity_financial_account](https://issues.civicrm.org/jira/browse/CRM-19357)

### CODE REVIEW

**Comment by Edzelopez** [comment]
Jenkins, please test this

**Comment by Edzelopez** [comment]
Jenkins, please test this

**Comment by Edzelopez** [comment]
Jenkins, please retest

**Comment by monishdeb** [comment]
jenkin test this please

**Comment by Edzelopez** [comment]
Is this okay to merge?

**Comment by monishdeb** [comment]
Tested on upgrade and install. The PR works for me. 

@eileenmcnaughton can you please merge this PR if the final patch looks good to you.

File: CRM/Upgrade/Incremental/php/FourSeven.php
Comment (colemanw) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: IMO this should be a generic function following the same pattern as `\CRM_Upgrade_Incremental_Base::addColumn()` and `\CRM_Upgrade_Incremental_Base::dropColumn`.

**Review by colemanw — DISMISSED** [review]

File: CRM/Upgrade/Incremental/php/FourSeven.php
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: @colemanw we've been recommending that function. I guess it could be moved to another class - but there are non-upgraded reasons to add & drop indexes.

My reservation about this & the 2 other open PRs for adding indexes is that on large sites adding indexes to large tables can be very slow and I feel like we should complete the work @aydun started to allow indexes to be declared & sites to reconcile them, rather than forcing them to either not upgrade to the next point version or have a significant outage.

Comment (colemanw) [inline CRM/Upgrade/Incremental/php/FourSeven.php]: I'm not saying don't use the `CRM_Core_BAO_SchemaHandler::createIndexes` function. I'm saying wrap it in a generic function in `CRM_Upgrade_Incremental_Base` instead of wrapping it in this one-off function.

**Comment by monishdeb** [comment]
@colemanw I have raised a separate ticket to add these two functions 
```
CRM_Upgrade_Incremental_Base::addIndex($ctx, $table, $column)
CRM_Upgrade_Incremental_Base::dropIndex($ctx, $table, $indexName)
```
So this PR is depedent on #10086 and I have also updated this PR accordingly.

**Comment by monishdeb** [comment]
Jenkin test this please

**Comment by monishdeb** [comment]
@colemanw I have made the changes like you suggested, can you merge it?

**Comment by eileenmcnaughton** [comment]
Note that if we merge this we should also merge the other PRs for adding indexes if 4.7.19 is going to be painful for large sites we should consolidate the pain

**Comment by monishdeb** [comment]
@eileenmcnaughton I have included the #10023 changes in this PR as it was dependent on this patch. Is there any other PR which adds index in 4.7.19 ? 

Bdw test build is failing for each PR at the moment

**Comment by monishdeb** [comment]
Jenkin test this please

**Comment by eileenmcnaughton** [comment]
I'm happy to merge the schema changes & changes for new installs. I would prefer to use a better approach for adding this & other schema changes via update - @aydun has done 95% of the work & #10109 has all except a small piece ready to go in.

Shall we get the DAO & schema changes merged & the new install sql for this & the other indexes & then worry about how to add them to existing installs in a non-damaging way?

**Comment by colemanw** [comment]
@monishdeb I merged #10131 can you please remove the similar code from here?

**Comment by monishdeb** [comment]
@colemanw @JoeMurray I am removing the upgrade code as suggested by Eileen [here](https://github.com/civicrm/civicrm-core/pull/10056#issuecomment-291841802). Will check what the best way to handle these indexes in upgrade after #10109 gets merged.

**Comment by eileenmcnaughton** [comment]
The test failures are related - likely either something wrong with the test or this has revealed an actual bug

**Comment by monishdeb** [comment]
@eileenmcnaughton the test build passed. Please have a look at the final patch

**Comment by eileenmcnaughton** [comment]
@monishdeb so 2 things here ... 

1) 'name' => 'index_entity_id_entity_table', - our standard naming is to include all the fields - I don't know if there is any issue with a longer name but if not we should use that
2) there will also be a change to the install sql to go in the commit

Note it would be ideal to add these indexes before running the mysql
civicrm_activitiy.activity_date_time https://issues.civicrm.org/jira/browse/CRM-20204
civicrm_mailing_event_queue.hash https://issues.civicrm.org/jira/browse/CRM-19383

**Comment by eileenmcnaughton** [comment]
Looks like the sql doesn't get committed - so that part is moot. We probably still should change that index name. Looks like 64 is the max http://stackoverflow.com/questions/16920001/maximum-length-of-a-column-name-in-mysql and I don't think we are there

**Comment by monishdeb** [comment]
@eileenmcnaughton I have changed the index name and then included the ```civicrm_generated.mysql``` after running UI. Ensured from UI that everything is running fine.
Here's the changed index in table 
1. Activity.status_id index 
image


2. Unique key ```index_entity_id_entity_table_account_relationship```
image

**Comment by eileenmcnaughton** [comment]
closed in favour of #10169

**Pull request closed without merging**
